src/pardibaal/DBM.cpp

```diff
@@ -354,53 +377,75 @@ namespace pardibaal {
     }
 
     void DBM::extrapolate_diagonal(const std::vector<val_t> &ceiling) {
+        const auto dim = this->dimension();
+
 #ifndef NEXCEPTIONS
-        if (this->dimension() != ceiling.size())
+        if (dim != ceiling.size())
             throw base_error("ERROR: Got max constants vector of size ", ceiling.size(), " but the DBM has ",
-                             this->dimension(), " clocks");
+                             dim, " clocks");
 #endif
-        DBM D(*this);
 
-        for (dim_t i = 0; i < D.dimension(); ++i) {
-            for (dim_t j = 0; j < D.dimension(); ++j) {
+        std::vector<std::pair<dim_t, dim_t>> modified_bounds;
+        modified_bounds.reserve(dim * (dim - 1));
+
+        std::vector<bool> close_zero(dim, false);
```

Comment:
There is a small benchmark to be had here if it is better to keep track of the dimensions that need the zero-close or have a bit-vector like this.
With the current construction you have to iterate over `dim` afterwords to check if it needs closing.

In the alternative construction you have
`std::vector<uint32_t> close_zero;` and `close_zero.push(j)` on line 407 instead.

Similar can be done for "skip" -- but here you want to keep the inverse.